app/src/containers/TestScreens/PopupTest.tsx

```diff
@@ -0,0 +1,48 @@
+import * as React from "react";
+import {ReactNode} from "react";
+import {Container, IContainerProps, IContainerState} from "../Container";
+import {PopupContainer} from "../PopupContainer";
+import {AddSongForm} from "../../components/AddSongForm";
+
+class PopupTest extends Container<IMainContainerProps, IMainContainerState> {
+
+    public static defaultProps: IMainContainerProps = {
+    };
+
+    protected constructor(props: IMainContainerProps) {
+        super(props);
+        this.state = {
+            popupOpen: false
+        };
+    }
+
+    openPopup = () => {
+        this.setState({popupOpen: true});
+    };
+
+    closePopup = () => {
+        this.setState({popupOpen: false});
+    };
+
+
+    public render(): ReactNode {
+        return (
+            <div className="main">
+                <button id="openPopup" onClick={this.openPopup}>Open Popup</button>
+                {this.state.popupOpen &&
```

Comment:
I wonder if it's better to put the PopUpModal on the Container Level so instead of doing this  you'll be writing a function that sets the contents of the popup modal, and the button callback that displays it. It makes sense in my head since there'll only be a max of 1 popup modal open at any time

Reply:
So, from my understanding, I move open/close Popup functions and the popupOpen state to the container abstract class. I would also have a new prop, `props.popupNode: ReactNode` for the container?

Reply:
yeah, make a protected function `displayPopUp` and `hidePopUp` or something in `Container`, and all you have to do in the screens that need it is to change the contents inside the popup (like in the constructor) and the button's onclick doesn't even need to be a new function because it's already implemented in the parent Container class

This is just my idea, feel free to implement it if you agree

Reply:
So, I am trying exactly that, but I can't import `PopupContainer` (which extends `Container`) in `Container`, so I'm not sure.

Reply:
`Container`
```tsx
protected popUpModalRef: PopUpModal;

protected showPopUp(callback?: () => void): void {
  /* show pop up */

  if (callback) {
    callback();
  }
}

protected hidePopUp(callback?: () => void): void {
  /* hide pop up */

  if (callback) {
    callback();
  }
}

render() {
  // save ref to modal
  <PopupModal ref={(ref) => {this.popUpModalRef = ref; }}/>
}
```

`PopUpModal`
```tsx
public setChild(childNode: ReactNode): void {
    this.setState({children: childNode});
}
```

Some Container that uses pop up modal
```tsx
constructor() {
  this.popUpModalRef.setChild(/* a new child node*/);
}

render() {
  // already implemented
  <button onAction={this.showPopUp}/>
}
```

It also works if you need multiple modals in the same page, you'd just switch out the child contents before you display it

Reply:
> So, I am trying exactly that, but I can't import `PopupContainer` (which extends `Container`) in `Container`, so I'm not sure.

Oh wtf, this should probably extend `EnhancedComponent` no?

Reply:
I don't think we should make Popup an enhanced component, because it is supposed to be a container, unless you think we should make it a component...? 
The way you did it is some next level JS that I only somewhat understand LOL. If you want, I can try it, otherwise let's keep it simple. I made some changes in the commit below.

Reply:
Also, just making sure that you are assuming that `PopupContainer` and `PopUpModal` are the same, because there is nothing called `PopUpModal` in my code

Reply:
Sorry yeah they're the same. Maybe something to keep in mind for the future then, I can try messing around with things

Reply:
Should I merge then?